File: app/src/main/java/cat/xojan/random1/domain/model/Podcast.java
package cat.xojan.random1.domain.model;

import android.os.Parcel;
import android.os.Parcelable;

import cat.xojan.random1.commons.ImageUtil;

public class Podcast implements Parcelable {

    private String mDescription;
    private String mFileUrl;
    private String mImageUrl;
    private String mProgram;
    private String mFilePath;
    private State mState;
    private int mImageDrawable;
    private String mProgramTitle;

    public Podcast(String description, String fileUrl, String program, String programTitle) {
        mDescription = description;
        mFileUrl = fileUrl;
        mImageUrl = ImageUtil.getPodcastImageUrl(program);
        mProgram = program;
        mState = State.LOADED;
        mImageDrawable = ImageUtil.getPodcastImageDrawable(program, programTitle);
        mProgramTitle = programTitle;
    }

    public Podcast(String program, String description, String filePath, State state,
                   String programTitle) {
        mProgram = program;
        mDescription = description;
        mFilePath = filePath;
        mImageUrl = ImageUtil.getPodcastImageUrl(program);
        mState = state;
        mImageDrawable = ImageUtil.getPodcastImageDrawable(program, programTitle);
    }

    public String getDescription() {
        return mDescription;
    }

    public String getFileUrl() {
        return mFileUrl;
    }

    public int getImageDrawable() {
        return mImageDrawable;
    }

    public String getProgram() {
        return mProgram;
    }

    public String getFilePath() {
        return mFilePath;
    }

    public State getState() {
        return mState;
    }

    public void setFilePath(String filePath) {
        mFilePath = filePath;
    }

    public void setState(State state) {
        mState = state;
    }

    public String getProgramTitle() {
        return mProgramTitle;
    }

    @Override
    public boolean equals(Object o) {
        if (o == this) return true;
        if (!(o instanceof Podcast)) {
            return false;
        }
        Podcast podcast = (Podcast) o;
        return mDescription.equals(podcast.mDescription) &&
                mProgram.equals(podcast.mProgram);
    }

    @Override
    public int hashCode() {
        int result = 17;
        result = 31 * result + mDescription.hashCode();
        result = 31 * result + mProgram.hashCode();
        return result;
    }

    protected Podcast(Parcel in) {
        mDescription = in.readString();
        mFileUrl = in.readString();
        mImageUrl = in.readString();
        mProgram = in.readString();
        mFilePath = in.readString();
        mState = (State) in.readValue(State.class.getClassLoader());
        mImageDrawable = in.readInt();
        mProgramTitle = in.readString();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mDescription);
        dest.writeString(mFileUrl);
        dest.writeString(mImageUrl);
        dest.writeString(mProgram);
        dest.writeString(mFilePath);
        dest.writeValue(mState);
        dest.writeInt(mImageDrawable);
        dest.writeString(mProgramTitle);
    }

    @SuppressWarnings("unused")
    public static final Parcelable.Creator<Podcast> CREATOR = new Parcelable.Creator<Podcast>() {
        @Override
        public Podcast createFromParcel(Parcel in) {
            return new Podcast(in);
        }

        @Override
        public Podcast[] newArray(int size) {
            return new Podcast[size];
        }
    };

    public enum State {
        LOADED,
        DOWNLOADING,
        DOWNLOADED
    }
}

File: app/src/main/java/cat/xojan/random1/domain/model/Podcast.kt
package cat.xojan.random1.domain.model

import android.os.Parcelable
import kotlinx.android.parcel.Parcelize
import java.util.*

@Parcelize
data class Podcast(var id: String,
                   var remoteUrl: String,
                   var filePath: String?,
                   var dateTime: Date?,
                   var durationSeconds: Long,
                   var programId: String,
                   var smallImageUrl: String?,
                   var bigImageUrl: String?,
                   var state: PodcastState,
                   var title: String,
                   var downloadReference: Long = 0
) : Parcelable {

    companion object {
        val PODCAST_STATE = "podcast_state"
        val PODCAST_FILE_PATH = "podcast_downloaded_file_path"
        val PODCAST_DOWNLOAD_REFERENCE = "podcast_download_reference"
        val PODCAST_PROGRAM_ID = "podcast_program_id"
        val PODCAST_BIG_IMAGE_URL = "podcast_big_image_url"
        val PODCAST_DURATION = "podcast_duration"
        val PODCAST_DATE = "podcast_date"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val podcast = other as Podcast

        if (remoteUrl != podcast.remoteUrl) return false
        if (programId != podcast.programId) return false
        return title == podcast.title
    }

    override fun hashCode(): Int {
        var result = remoteUrl.hashCode()
        result = 31 * result + programId.hashCode()
        result = 31 * result + title.hashCode()
        return result
    }
}